drgn/cli.py

```diff
@@ -198,75 +209,125 @@ def main() -> None:
         if missing_debug_info_warning is not None:
             print(missing_debug_info_warning, file=sys.stderr)
+        init_globals: Dict[str, Any] = {"prog": prog}
         runpy.run_path(script, init_globals=init_globals, run_name="__main__")
     else:
-        sys.path.insert(0, "")
-
-        import atexit
-        import readline
-
-        from drgn.internal.rlcompleter import Completer
-
-        init_globals["drgn"] = drgn
-        drgn_globals = [
-            "NULL",
-            "Object",
-            "cast",
-            "container_of",
-            "execscript",
-            "offsetof",
-            "reinterpret",
-            "sizeof",
-        ]
-        for attr in drgn_globals:
-            init_globals[attr] = getattr(drgn, attr)
-        init_globals["__name__"] = "__main__"
-        init_globals["__doc__"] = None
-
-        histfile = os.path.expanduser("~/.drgn_history")
-        try:
-            readline.read_history_file(histfile)
-        except OSError as e:
-            if not isinstance(e, FileNotFoundError) and not args.quiet:
-                print("could not read history:", str(e), file=sys.stderr)
 
-        def write_history_file() -> None:
-            try:
-                readline.write_history_file(histfile)
-            except OSError as e:
-                if not args.quiet:
-                    print("could not write history:", str(e), file=sys.stderr)
+        def banner_func(banner: str) -> str:
+            if missing_debug_info_warning is not None:
+                return f"{banner}\n{missing_debug_info_warning}"
+            else:
+                return banner
+
+        run_interactive(prog, banner_func=banner_func, quiet=args.quiet)
+
+
+def run_interactive(
+    prog: drgn.Program,
+    banner_func: Optional[Callable[[str], str]] = None,
+    globals_func: Optional[Callable[[Dict[str, Any]], Dict[str, Any]]] = None,
+    quiet: bool = False,
+) -> None:
+    """
+    Run the drgn interactive command line interface.
+
+    This function allows your application to embed the same REPL drgn provides
+    when it is run on the command line in interactive mode. Simply provide an
+    already configured :class:`drgn.Program`, and optionally customize the
+    banner and/or globals, and this function will create the REPL and run it
+    until the user exits.
+
+    If you are using this function to embed within another application using
+    readline, then you'll need to take special care. While drgn tries to clean
+    up any modifications to global state made during this function, it cannot
+    save and restore the readline settings. In particular, drgn clears and uses
+    the readline history, and it binds a custom completer to the TAB key. You
+    should be sure to save your application's history and clear any custom
+    keybindings or completers prior to calling this function, and then re-read
+    your application's history and restore keybindings and completion functions
+    on return.
```

Comment:
Let's put this in a Sphinx [note](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html).